src/core/public/plugins/plugins_service.ts

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { CoreSetup } from '..';
+import { CoreService } from '../../types';
+import { CoreContext } from '../core_system';
+import { Plugin } from './plugin';
+import { createPluginInitializerContext, createPluginSetupContext } from './plugin_context';
+
+/** @internal */
+export type PluginsServiceSetupDeps = CoreSetup;
+
+/** @internal */
+export interface PluginsServiceSetup {
+  contracts: Map<string, unknown>;
+}
+
+/**
+ * Service responsible for loading plugin bundles, initializing plugins, and managing the lifecycle
+ * of all plugins.
+ *
+ * @internal
+ */
+// tslint:disable max-classes-per-file
+export class PluginsService implements CoreService<PluginsServiceSetup> {
+  private readonly plugins: Map<string, Plugin<unknown, Record<string, unknown>>> = new Map();
+
+  constructor(private readonly coreContext: CoreContext) {}
+
+  public async setup(deps: PluginsServiceSetupDeps) {
+    // Construct plugin wrappers
+    deps.injectedMetadata
+      .getPlugins()
+      .forEach(({ id, plugin }) =>
+        this.plugins.set(id, new Plugin(plugin, createPluginInitializerContext(deps, plugin)))
+      );
+
+    // Load plugin bundles
+    await this.loadPluginBundles(deps.basePath.addToPath);
+
+    // Setup each plugin with correct dependencies
+    const contracts = new Map<string, unknown>();
+    for (const [pluginName, plugin] of this.plugins.entries()) {
```

Comment:
I disagree on this slightly.

> otherwise it's a bit confusing why we read (get) before the very first write (set)

`this.plugins` is populated in the very first line of this function, not below (`contracts` is populated below). It's part of the PluginsService's structure that `this.plugins` is in topological order. I'll add that as a TSDoc comment to the private variable, but I don't think it's necessary to rename it here.